헤더, 서치바 구현 #10

joojjang opened the pull request:
## **PR**

### ✨ 작업 내용

- #10 헤더, 서치바 만들었습니다. 
- 아직 아이콘들에 페이지 라우팅은 안 했습니다. 
- 서치바에는 검색어 삭제까지는 동작 구현했습니다. 

### 📌 참고 사항

- vite-plugin-svgr 깔아서 설정 파일 건드리니까 컴포넌트로 import 되더라고요. 다만 임포트 문을  
import 컴포넌트명 from "@/.../파일명.svg?react";  
이렇게 파일명 뒤에 ?react를 붙여야 합니다.  
아이콘 매번 임포트 하는 게 귀찮고 코드도 많이 다를 것 같아서 추후 아이콘 스타일 파일도 하나 만들지 생각 중... 만든다면 빨리 만들어야 할 것 같습니다.  
- 아이콘 이름을 하나로 정해서 favorite이면 favorite, heart면 heart, like면 like ... 하나로 정해야할 것 같습니다.  
그리고 아이콘이 버튼으로 쓰인다면 태그를 button으로 하는 게 html 구조상 맞긴 한데...

### ✏ Git Close
close #10

joojjang committed:
feat(header): 헤더 컴포넌트 구현

joojjang committed:
fix(header): leftSideChildren을 선택 prop으로 지정

joojjang committed:
fix(svgr): svg를 컴포넌트로 인식 못하는 오류 수정 중

- vite-plugin-svgr 설치 후 설정 파일을 커스텀했으나 아직 미해결

joojjang committed:
feat(SearchBar): 서치바 공통 컴포넌트 구현

- 아이콘 미삽입

joojjang committed:
fix(Header): 헤더 너비 수정 100vw -> 100%

joojjang committed:
feat(SearchBar): 아이콘 svg를 컴포넌트로 import 및 삽입 / favorite 유무 상태, cancel 버튼 유무 상태 구분

- favorite 유무에 따라 인풋 요소 너비가 달라져서 InputBox 만들고 그 기준으로 아이콘 absolute로 위치시킴
- 새 svg 파일 추가
- includeFavorite와 searchWord 상태 구분하여 동적으로 요소 렌더링

joojjang committed:
feat(SearchBar): 검색어 지우기 구현

- onClick 핸들러 함수 추가

joojjang committed:
design(SearchBar): 폰트 사이즈, 마진 조정

joojjang committed:
design(Header): 여백, 폰트 사이즈 등 조정

seung365 commented on src/components/Header/index.tsx:
```diff
@@ -0,0 +1,71 @@
+      return (
+        <>
+          <IconBox>{leftSideChildren}</IconBox>
+          <TitleBox>{title}</TitleBox> {/* 있을 때만 렌더링됨 */}
```
이 부분 있을때만 렌더링 된다면 삼항연산자나 논리연산자를 사용하는게 어떨까요

joojjang replied:
어차피 옵셔널 속성이라 삼항연산자 굳이 사용 안 해도 될 거라고 생각했습니당
이해를 돕기 위해 주석만 추가한 거예요

seung365 commented on src/components/Header/index.tsx:
```diff
@@ -0,0 +1,71 @@
+export default Header;
+
+// styles
+const HEADER_HEIGHT = '44px';
```
rem 대신 px을 쓰신 이유가 있을까요?

joojjang replied:
고정 너비라서 그냥 그렇게 지정했었는데 안에 글자가 들어갈 수도 있으니 바꾸겠습니다~

seung365 commented on src/components/Header/index.tsx:
```diff
@@ -0,0 +1,71 @@
+        <>
+          <IconBox>{leftSideChildren}</IconBox>
+          <TitleBox>{title}</TitleBox> {/* 있을 때만 렌더링됨 */}
+          <IconBox>{rightSideChildren}</IconBox> {/* 있을 때만 렌더링됨 */}
```
그렇게 되면 주석 지워주셔도 좋을꺼 같습니다

seung365 commented on src/components/Header/index.tsx:
```diff
@@ -0,0 +1,71 @@
+
+const TitleBox = styled.div`
+  position: absolute;
+  // 중앙 배치
```
요 주석도 없어도 될꺼 같습니다

seung365 commented on src/components/SearchBar/index.tsx:
```diff
@@ -1,24 +1,37 @@
+          value={searchWord}
+          onChange={(e) => {
+            setSearchWord(e.target.value);
+            //   console.log(e.target.value); // test
```
이런 테스트 후의 주석도 없어도 될꺼 같아요

seung365 commented on src/components/SearchBar/index.tsx:
```diff
@@ -29,7 +29,14 @@ const SearchBar = ({ includeFavorite = false }: SearchBarProps) => {
-        {searchWord.trim().length > 0 && <StyledCancelIcon />}
+        {searchWord.trim().length > 0 && (
+          <StyledCancelIcon
+            onClick={(e) => {
```
바로 이벤트핸들러 안에 넣는것도 좋은데 두줄 이상이면 함수로 따로 구현해서 넣으면 더 좋을꺼 같아요

joojjang replied:
재사용할 것 같지 않아서 바로 넣었는데 
가독성을 위해 그렇게 하겠습니다~

joojjang committed:
fix(Header): 높이 단위 rem으로 변경

joojjang committed:
remove(SearchBar): 테스트 주석 삭제

joojjang committed:
refactor(SearchBar): 함수 분리

joojjang committed:
refactor(IconButton): 아이콘 버튼 스타일 구현, reset.css에서 버튼 스타일 변경

- IconButton으로 버튼인 아이콘들 감쌈
- 전역 스타일로 버튼에 outline: none 추가

joojjang committed:
refactor(IconButton): 컴포넌트화

- 재사용을 위해 추가되어야하는 prop이 많다면 굳이 재사용 X (CancelIconButton 같은 것)
- 아이콘이지만 버튼이 아니라면 IconButton 사용 X

joojjang commented:
추가된 사항:  
아이콘도 리팩토링 해야 할 것 같다고 했었는데  
IconButton 컴포넌트 만들어서 버튼으로 기능하는 아이콘은 재사용할 수 있게 만들었습니다. 태그도 button이에요!
버튼이 아니거나 재사용을 위해 굳이 다른 여러 prop 집어넣어야 할 경우에는 그냥 재사용 안 했습니다. (SearchBar에서 CancelIconButton)

joojjang committed:
fix(SearchBar): 아이콘 사이즈 단위 px -> rem 변경

seung365 commented:
수고하셨습니다!!
